File: H-Index - Leetcode 274/H-Index - Leetcode 274.py
class Solution:
    def hIndex(self, citations: List[int]) -> int:
        n = len(citations)
        paper_counts = [0] * (n+1)

        for c in citations:
            paper_counts[min(n, c)] += 1
        
        h = n
        papers = paper_counts[n]
        
        while papers < h:
            h -= 1
            papers += paper_counts[h]
        
        return h
        # Time: O(n), Space: O(n)

File: H-Index - Leetcode 274/H-Index - Leetcode 274.cpp
#include <vector>
#include <algorithm>

class Solution {
public:
    int hIndex(std::vector<int>& citations) {
        int n = citations.size();
        std::vector<int> paperCounts(n + 1, 0);

        for (int c : citations) {
            paperCounts[std::min(n, c)]++;
        }

        int h = n;
        int papers = paperCounts[n];

        while (papers < h) {
            h--;
            papers += paperCounts[h];
        }

        return h;
    }
};
